### ISSUE TITLE
Add hardware accelerator

### ISSUE THREAD

**Issue by x94carbone** [autonomio/talos#55]
Obviously this is something that could be done later on since it is not exactly something I'd expect the average user to use. However, tensorflow/Keras makes it _very easy_ to implement parallel GPU training when available. I already have the code to do this cleanly. Its just a matter of how (and if) to implement this.

In principle this can only be a good thing since tensorflow/Keras `multi_gpu_model` is essentially straightforward (I struggled with it for so long 🙃) if you know what you're doing with it. The only level of parallelism is at the batch level, so everything in Talos could work on top of it.

@mikkokotila thoughts on this? Google.colab already accelerates any training I've seen done with Keras massively without any changes to code whatsoever, but to my knowledge you only get one GPU at a time. People with access to more could experience up to an 8-fold decrease in training time with a feature like this.

**Comment by mikkokotila** [comment]
I think this is a really great and important addition. Especially if you have already figured out :) I could take an arrayed Tesla instance from AWS to test it out!

**Comment by Manofdesign** [comment]
multi_gpu_model in the build_model function using the Keras functional api seems to work just fine in talos.Scan when experimental_functional_support=True.

I got a box with 2 cards and both light up when running, so I thought talos already supported this feature.  Is there some other aspect on this that I am totally missing?

**Comment by Manofdesign** [comment]
@x94carbone thanks for clarifying.  It definitely works.  I monitor gpu load, gpu memory usage, temp and fan speed on both cards and all measures indicate both are in use.  Multi GPU support is the main feature which lead me to talos.  

I was using GridSearch prior, but it seemed multi-gpu on gridsearch failed for me.  I suspect the memory leak issue.  I saw the git version of talos had the method to reset the session between model builds to clear GPU memory.  That worked great in talos.  When I have it sent to false I can cause a memory leak error from doing enough runs in talos.  When I set it to true, talos clears the memory each run.  Confirmed by setting allow growth in tensorflow session to true and the memory clears and grows to needed amount before each run starts on both GPUs.

**Comment by mikkokotila** [comment]
@Manofdesign 

The "allow growth" feature that you mention, is it this one:

gpu_options = tf.GPUOptions(per_process_gpu_memory_fraction=0.333)
sess = tf.Session(config=tf.ConfigProto(gpu_options=gpu_options))

This seems like something we need to add as an option for the user directly from Scan(). I will test next few days and add it.

**Comment by mikkokotila** [comment]
Would this do it: 

```
import tensorflow as tf
from keras.backend.tensorflow_backend import set_session


def max_gpu_memory(fraction):

    '''Sets the max used memory as a fraction for tensorflow
    backend

    fraction :: a float value (e.g. 0.5 means 4gb out of 8gb)

    '''

    config = tf.ConfigProto()
    config.gpu_options.per_process_gpu_memory_fraction = fraction
    set_session(tf.Session(config=config))
```

...and we call this in the beginning of the experiment, or if clear_tf_session is set to True, then we do it for every permutation of the experiment? 

I'm creating tf_config.py under /utils where similar things would go in the future. So this would be imported as:

`from talos.utils.tf_config import max_gpu_memory`

**Comment by Manofdesign** [comment]
@mikkokotila Here is the syntax that I used.  This assumes import keras.backend as K called out earlier.  
`
gpu_options = K.tf.GPUOptions(allow_growth=True)
K.set_session(K.tf.Session(config=tf.ConfigProto(gpu_options=gpu_options)))

h = talos.Scan(x=x_train, y=y_train,
               x_val=x_val, y_val=y_val,
               shuffle=False,
               params=p,
               dataset_name='Simple_MLP_EmbedFalse_featscaleneg',
               experiment_no='2',
               model=create_model,
               grid_downsample=.01,
               last_epoch_value=True,
               experimental_functional_support=True, 
               clear_tf_session=True #very important to prevent memory leak and scan failing
               )`


To include the fixed percentage

`gpu_options = K.tf.GPUOptions(allow_growth=True, per_process_gpu_memory_fraction=0.5)`

As long as clear_tf_session=True is set, it clears the memory on each permutation but remembers the session settings when it starts the next, so I don't think it needs to be inside the permutation loop.

This brings another thought.  The nature of my current experiment only utilizes < 50% on both cards.  GridSearch has an n_Jobs option where you can set the total number of parallel permutations to run at the same time.  I hate seeing idle compute capacity sitting there and want to squeeze every drop out possible.  Any way to config talos to run parallel permutations.  In that case I think there would need to be more than one session set for each parallel permutation to run in.

I have no idea if this is even possible.  I can do an initial test by setting the per_process_gpu_memory_fraction=0.5 and running two terminal sessions at once to see what happens.  My hope is that both will be running and GPUs will be near 100% usage.

**Comment by Manofdesign** [comment]
Confirmed.  I can run 2 separate talos.scan sessions with

```
gpu_options = K.tf.GPUOptions(allow_growth=True, per_process_gpu_memory_fraction=0.5)
K.set_session(K.tf.Session(config=tf.ConfigProto(gpu_options=gpu_options)))
```

and both run in parallel nearly utilizing full GPU capacity.  

It would be awesome if talos.Scan had an N_parallel_jobs option where talos would enable multiple sessions as called out and would feed permutations to the session instance that is ready next.

**Comment by mikkokotila** [comment]
Thanks @Manofdesign and I agree, this is something we need to take seriously. Would this be something you are comfortable contributing with?

**Comment by mikkokotila** [comment]
@Manofdesign @x94carbone 

As I understand this will be probably the most important think to get right in the roadmap, I'll take a step back and try to dissect. 

It seems that here we are talking now about bigger picture of resources use. First, we have two kinds of systems:

- CPU based
- GPU based 

This is quite important, as for some prediction tasks (e.g. batch_size tends to be small) even a low-end CPU can be an order of magnitude faster than high-end GPU based system. 

Then we have three kinds of distributions: 

- a single resource machine (e.g. a single machine with a single GPU)
- a multi resource machine (e.g. a single machine with many GPUs) 
- multi-machine, single-resources (e.g. many machines, each with single GPU) 
- multi-machine, multi-resource (e.g. many machines, each with many GPUs) 

I'm not sure if multi-CPU utilization is as effective out-of-the-box as CPU core utilization is with Keras/TensorFlow, or is it more like multi-GPU case. 

In terms of experiments, we have several scenarios: 

- a single experiment run as a single job 
- a single experiment run as multiple jobs
- multiple experiments run as a single job 
- multiple experiments run as multiple jobs

I appreciate the fact that multiple experiments run as a single job might not be relevant, but I'm including it for the sake of ontological completeness regarding this topic. 

Basically we are left with 2x4x3=12 relevant cases to figure out and make sure that we cover each in a meaningful way. I think the most complex will be "distributed" scenario where we have multiple machines with multiple GPUs. 

The other interesting thing here might be (at least for very long runs) to have a switch for automatically choosing a given system if CPU, GPU (or TPU) should be used. For example, I have a case where on Google colab it takes 10s per permutation for CPU, 70s/p with GPU and 40s per permutation with TPU. So it would be very easy for the system to auto-optimize and pick CPU for the experiment without the user having to worry about it.

Thoughts?

**Comment by Manofdesign** [comment]
@mikkokotila @x94carbone Will try and respond to multiple comments in this one.  Regarding contributing.  I would be very interested in contributing.  The disclaimer I have to include is that I have <1 year experience in Python and Keras and will likely need some guidance and mentor-ship in some areas such as proper code structure and Keras.    

I like the idea of covering various scenarios.  I do wonder of the scope should be limited.  I think prioritizing multi GPU local and distributed GPU (multi machine X single or multi GPU) while allowing multiple jobs in parallel would provide the most bang for effort buck.  Especially if it could be easily integrated with Horovod.  

I have never ran into a Keras model where the CPU performed better than the GPU.  Would be interested in the details of the Google colab scenario where the CPU was out performing GPUs from a learning perspective.  My experiences are on a local machine that has either a 1050ti or two 1060s.  Typically I see my GPU performs 10x faster than my CPU.  When running a GPU enabled session on Google Colab, it seems on par with my local 1050ti.  

My vote would be to prioritize the GPU scenarios first and if the community really screams for distributed CPU functionality, tackle it then.  

In order to auto optimize wouldn't the system have to run a set of permutations on each compute scenario to see which is best?  Then we would have to be confident that sampling of permutations was representative enough of the complete set that the best compute scenario was chosen.  Trying to wrap my head around the workflow that would always lead to the best auto selected choice.    

By the time I am ready to run parameter sweeps on a model I feel that I already know the compute scenario I want to use based on initial testing.  Just being able to manually select the scenario would be sweet.  

Back on contributing.  What do you suggest as the first step?

**Comment by Manofdesign** [comment]
@mikkokotala I would be happy to test those scenarios.  My multi-GPU box is tied up for a few days running a big experiment, but can run the force_cpu, single job-single-gpu, multiple job-single-gpu scenarios while that is going.  I am working to get Horovod installed and working on multiple machines to work with Keras.  One that is up and running I can make an attempt to look at the Horovod implemetnation.  As far as providing code, should I create a separate branch that can be merged later down the road?

**Comment by mikkokotila** [comment]
@Manofdesign wonderful, and yes, you could make your own for, and then make PR to daily-dev. 

In the meantime, I got myself one multi-GPU system I'm setting up now, in the next 2-3 weeks should be able to have another system (once 2080 Ti is available) which will also have a very high end CPU so it will be a good platform to test various scenarios CPU vs. GPU (as on a system like that the GPU should theoretically speaking never be bottle necked by CPU).

**Comment by mikkokotila** [comment]
I ended up spending 2 weeks just on benchmarking to get to the bottom of the matters. I have one system with 2080Ti, one system with 2x1070Ti and then I switched the cards between the machines (so different CPU) and did with both overclocked CPU and not. The results ended up being a lot so it will take time to get it out in some meaningful way. In the meantime, generally speaking CPU is always better than GPU for smaller models with smaller batch sizes. The other important insight is that generally speaking 2x a given GPU needs a special case before you get 2x the compute power. In fact, in many cases 2x 1070Ti is worse than 1x 1070Ti. Also sometimes 1x 1070Ti is better than 2080Ti (on the same CPU). Some of the more consistent things I found in the results was that using a tensorflow optimizer inside TFOptimizer wrapper is much faster than same optimizer as Keras optimizer (even both are used in Keras the same way). The other thing was to either compile TF from source, or use Conda instead of pip. For LSTM model, obviously, CudnnLSTM is 10-20x faster than LSTM, so that was obviously the single biggest factor decreasing train time
